OMCompiler/Compiler/NBackEnd/NBEquation.mo

```diff
@@ -170,6 +170,92 @@ public
         else EQ_ATTR_DEFAULT_UNKNOWN;
       end match;
     end getAttributes;
+
+    // TODO! use referenceEq when updating stuff
+    function map
+      "Traverses all expressions of the equations and applies a function to it.
+      Optional second input to also traverse crefs, only needed for simple
+      eqns, when eqns and algorithms."
+      input output Equation eq;
+      input MapFuncExp funcExp;
+      input Option<MapFuncCref> funcCrefOpt = NONE();
+      partial function MapFuncExp
+        input output Expression e;
+      end MapFuncExp;
+      partial function MapFuncCref
+        input output ComponentRef c;
+      end MapFuncCref;
+    algorithm
+      eq := match eq
+        local
+          Equation qualEq, body;
+          MapFuncCref funcCref;
+          list<list<Equation>> eqnstrue = {};
+
+        case qualEq as SCALAR_EQUATION()
+          algorithm
+            qualEq.lhs := funcExp(qualEq.lhs);
+            qualEq.rhs := funcExp(qualEq.rhs);
+        then qualEq;
+
+        case qualEq as ARRAY_EQUATION()
+          algorithm
+            qualEq.lhs := funcExp(qualEq.lhs);
+            qualEq.rhs := funcExp(qualEq.rhs);
+        then qualEq;
+
+        case qualEq as SIMPLE_EQUATION() guard(isSome(funcCrefOpt))
+          algorithm
+            SOME(funcCref) := funcCrefOpt;
+            qualEq.lhs := funcCref(qualEq.lhs);
+            qualEq.rhs := funcCref(qualEq.rhs);
+        then qualEq;
+
+        case qualEq as RECORD_EQUATION()
+          algorithm
+            qualEq.lhs := funcExp(qualEq.lhs);
+            qualEq.rhs := funcExp(qualEq.rhs);
+        then qualEq;
+
+        case qualEq as ALGORITHM()
+          algorithm
+            qualEq.alg := Algorithm.mapExp(qualEq.alg, funcExp);
+            if isSome(funcCrefOpt) then
+              SOME(funcCref) := funcCrefOpt;
+              qualEq.inputs := List.map(qualEq.inputs, funcCref);
+              qualEq.outputs := List.map(qualEq.outputs, funcCref);
+            end if;
+        then qualEq;
+
+        case qualEq as IF_EQUATION()
+          algorithm
+            qualEq.conditions := List.map(qualEq.conditions, funcExp);
+            for eqn_lst in qualEq.eqnstrue loop
+              eqnstrue := List.map2(eqn_lst, map, funcExp, funcCrefOpt) :: eqnstrue;
+            end for;
+            qualEq.eqnstrue := listReverse(eqnstrue);
+            qualEq.eqnsfalse := List.map2(qualEq.eqnsfalse, map, funcExp, funcCrefOpt);
+        then qualEq;
+
+        case qualEq as FOR_EQUATION()
+          algorithm
+            qualEq.range := funcExp(qualEq.range);
+            qualEq.body := map(qualEq.body, funcExp, funcCrefOpt);
+            qualEq.range := funcExp(qualEq.range);
```

Comment:
No need to apply funcExp to the range twice, right?

Reply:
Yea totally correct. Copy and paste error :D